core/src/main/java/com/sequenceiq/cloudbreak/service/multiaz/MultiAzCalculatorService.java

```diff
@@ -248,4 +255,118 @@ private boolean isPreferredSubnetsSpecifiedForScaling(NetworkScaleDetails networ
         return networkScaleDetails != null
                 && CollectionUtils.isNotEmpty(networkScaleDetails.getPreferredSubnetIds());
     }
+
+    public Set<GroupPlacement> prepareGroupAzMap(Map<String, String> subnetAzPairs, Set<InstanceGroup> instanceGroups) {
+        Set<String> groups = getGroups(instanceGroups);
+        Set<String> availabilityZones = getAvailabilityZones(subnetAzPairs);
+
+        if (availabilityZones.size() < NUMBER_OF_GROUPS || groups.size() != NUMBER_OF_GROUPS) {
+            return Collections.emptySet();
+        }
+
+        Set<GroupPlacement> groupPlacements = new HashSet<>();
+        Iterator<String> availabilityZonesIt = availabilityZones.iterator();
+        groups.forEach(gp -> {
+            String availabilityZone = availabilityZonesIt.next();
+            GroupPlacement groupPlacement = new GroupPlacement(gp, availabilityZone);
+            groupPlacements.add(groupPlacement);
+
+            subnetAzPairs
+                    .entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue().equals(availabilityZone))
+                    .map(Entry::getKey)
+                    .forEach(subnet -> groupPlacement.addSubnet(subnet));
+        });
+        return groupPlacements;
+    }
+
+    private Set<String> getAvailabilityZones(Map<String, String> subnetAzPairs) {
+        return subnetAzPairs
+                .entrySet()
+                .stream()
+                .map(Entry::getValue)
+                .collect(Collectors.toSet());
+    }
+
+    private Set<String> getGroups(Set<InstanceGroup> instanceGroups) {
+        return instanceGroups
+                .stream()
+                .filter(ig -> Objects.nonNull(ig.getHints()))
+                .map(InstanceGroup::getHints)
+                .flatMap(Set::stream)
+                .collect(Collectors.toSet());
+    }
+
+    public void calculate(Map<String, String> subnetAzPairs, InstanceGroup instanceGroup, Set<GroupPlacement> groupPlacements,
+            SdxClusterShape sdxClusterShape) {
+        if (isGroupMultiAZ(sdxClusterShape, groupPlacements)) {
+            calculateByGroup(instanceGroup, groupPlacements);
+        } else {
+            calculateByRoundRobin(subnetAzPairs, instanceGroup);
+        }
+    }
+
+    private boolean isGroupMultiAZ(SdxClusterShape sdxClusterShape, Set<GroupPlacement> groupPlacements) {
+        return !groupPlacements.isEmpty() &&
+                sdxClusterShape == SdxClusterShape.SCALABLE;
+    }
+
+    private void calculateByGroup(InstanceGroup instanceGroup, Set<GroupPlacement> groupPlacementList) {
```

Comment:
Isn't an equivalent method for the below not needed?

    public void calculateByRoundRobin(Map<String, String> subnetAzPairs, InstanceGroupDto instanceGroupDto, InstanceMetaData instanceMetaData,
            NetworkScaleDetails networkScaleDetails) {

Reply:
The equivalent method is required for upscale. It will be handled by another ticket.